extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/ControllerImpl.java

```diff
@@ -0,0 +1,2040 @@
+      final SortColumn column = clusterByColumns.get(i);
+      final List<String> outputColumns = columnMappings.getOutputColumnsForQueryColumn(column.columnName());
+
+      // DimensionRangeShardSpec only handles ascending order.
+      if (column.descending()) {
+        return Collections.emptyList();
+      }
+
+      ColumnType columnType = signature.getColumnType(column.columnName()).orElse(null);
+
+      // DimensionRangeShardSpec only handles strings.
+      if (!(ColumnType.STRING.equals(columnType))) {
+        return Collections.emptyList();
+      }
+
+      // DimensionRangeShardSpec only handles columns that appear as-is in the output.
+      if (outputColumns.isEmpty()) {
+        return Collections.emptyList();
+      }
+
+      shardColumns.add(outputColumns.get(0));
+    }
+
+    return shardColumns;
+  }
+
+  private static boolean isClusterByBoosted(final ClusterBy clusterBy)
+  {
+    return !clusterBy.getColumns().isEmpty()
+           && clusterBy.getColumns()
+                       .get(clusterBy.getColumns().size() - 1)
+                       .columnName()
+                       .equals(QueryKitUtils.PARTITION_BOOST_COLUMN);
+  }
+
+  private static StringTuple makeStringTuple(
+      final ClusterBy clusterBy,
+      final RowKeyReader keyReader,
+      final RowKey key
+  )
+  {
+    final String[] array = new String[clusterBy.getColumns().size() - clusterBy.getBucketByCount()];
+    final boolean boosted = isClusterByBoosted(clusterBy);
+
+    for (int i = 0; i < array.length; i++) {
+      final Object val = keyReader.read(key, clusterBy.getBucketByCount() + i);
+
+      if (i == array.length - 1 && boosted) {
+        // Boost column
+        //noinspection RedundantCast: false alarm; the cast is necessary
+        array[i] = StringUtils.format("%016d", (long) val);
+      } else {
+        array[i] = (String) val;
+      }
+    }
+
+    return new StringTuple(array);
+  }
+
+  private static Pair<List<DimensionSchema>, List<AggregatorFactory>> makeDimensionsAndAggregatorsForIngestion(
+      final RowSignature querySignature,
+      final ClusterBy queryClusterBy,
+      final List<String> segmentSortOrder,
+      final ColumnMappings columnMappings,
+      final boolean isRollupQuery,
+      final Query<?> query
+  )
+  {
+    final List<DimensionSchema> dimensions = new ArrayList<>();
+    final List<AggregatorFactory> aggregators = new ArrayList<>();
+
+    // During ingestion, segment sort order is determined by the order of fields in the DimensionsSchema. We want
+    // this to match user intent as dictated by the declared segment sort order and CLUSTERED BY, so add things in
+    // that order.
+
+    // Start with segmentSortOrder.
+    final Set<String> outputColumnsInOrder = new LinkedHashSet<>(segmentSortOrder);
+
+    // Then the query-level CLUSTERED BY.
+    // Note: this doesn't work when CLUSTERED BY specifies an expression that is not being selected.
```

Comment:
Does the SQL layer prevent this state of affairs? The case only occurs in tasks submitted outside of SQL?

Reply:
IIRC, the SQL layer allows it. In this case the expression is still used for partitioning data between segments, but not for sorting data within a segment. In the catalog, it may make sense to expose a "segment sort by" for advanced users, to make this case clearer. (Today, CLUSTERED BY covers both partitioning and sorting, because 95% of the time people want the same key for both. So it would really be a 5%-of-the-time advanced users thing.)

Reply:
Extended the comment to spell this out:

```
    // Then the query-level CLUSTERED BY.
    // Note: this doesn't work when CLUSTERED BY specifies an expression that is not being selected.
    // Such fields in CLUSTERED BY still control partitioning as expected, but do not affect sort order of rows
    // within an individual segment.
```